### PR TITLE
Path splitting

### PR DESCRIPTION
**Pull request by robdockins**
This PR implements an execution feature for path splitting, and a variety of refactorings and fixes to make it work properly.  It also fixes an issue with the bounded execution feature (CF #185) by recording the relevant data structures in a crucible global variable instead of alongside the simulator in an IORef.

The profiling simulator feature exhibits a similar problem, but it is less clear how to fix it.  As such, this PR is probably not ready to merge yet.

### CODE REVIEW

**Comment by robdockins** [comment]
I've now piped through options for changing the `path-strategy` option, which can be either `always-merge` (the previous behavior) or `split-dfs` which performs path splitting and exploration in depth-first manner.

Rather than solve the problems that arise from the interaction of path splitting and profiling (which seem rather difficult), I've made it so that the profiling option is incompatible with path splitting in crux. At a later time, we can revisit this.

File: crucible/src/Lang/Crucible/Simulator/EvalStmt.hs [hunk 1]
```diff
@@ -551,6 +552,40 @@ singleStepCrucible verb exst =
...
+  ExecutionFeatureModifiedState ::
+     ExecState p sym ext rtp -> ExecutionFeatureResult p sym ext rtp
+
+  -- | This exeuction feature result indicates that the state was modified
+  --   in an essential way that transforms it into new state altogether.
+  --   When this result is returned, it preempts any later execution
+  --   features and the main simulator loop and instead returns to the head
+  --   of the execution feature stack.
+  --
+  --   NOTE: In particular, the exeuction feature will encounter the
```
Comment (atomb) [inline crucible/src/Lang/Crucible/Simulator/EvalStmt.hs]: Minor typo: `exeuction`.

Comment (atomb) [reply crucible/src/Lang/Crucible/Simulator/EvalStmt.hs]: Also the same typo below.

File: crucible/src/Lang/Crucible/Simulator/EvalStmt.hs [hunk 2]
```diff
@@ -551,6 +552,40 @@ singleStepCrucible verb exst =
...
+  --   state again before the simulator loop.  It is therfore very
+  --   important that the exeuction feature be prepared to immediately
+  --   encounter the same state again and make significant execution
+  --   progress on it, or ignore it so it makes it to the main simulator
+  --   loop.  Otherwise, the execution feature will loop back to itself
+  --   infinetely, starving out useful work.
```
Comment (atomb) [inline crucible/src/Lang/Crucible/Simulator/EvalStmt.hs]: Typo: `infinetely`.

File: crucible/src/Lang/Crucible/Simulator/PathSplitting.hs [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+  IO (ExecState p sym ext rtp)
+restoreWorkItem (WorkItem branchPred loc frm st assumes) =
+  do let sym = st ^. stateSymInterface
+     setCurrentProgramLoc sym loc
+     restoreAssumptionState sym assumes
+     addAssumption sym (LabeledPred branchPred (ExploringAPath loc (pausedLoc frm)))
+     let ctx = st ^. stateTree . actContext
+     runReaderT (resumeFrame frm ctx) st
+
+{-
```
Comment (atomb) [inline crucible/src/Lang/Crucible/Simulator/PathSplitting.hs]: Is there any reason to keep this commented-out code?

File: crux/src/Crux/Log.hs [hunk 4]
```diff
@@ -59,6 +59,9 @@ outputColored c msg =
 sayOK :: Logs => String -> String -> IO ()
 sayOK = sayCol Green
 
+sayWarn :: Logs => String -> String -> IO ()
```
Comment (atomb) [inline crux/src/Crux/Log.hs]: Good addition.

**Pull request merged**